filemanager.go

```diff
@@ -0,0 +1,358 @@
+// Copyright 2023 qbee.io
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package client
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+)
+
+// Manager manages the sync operation.
+type Manager struct {
+	client      *Client
+	del         bool
+	dryrun      bool
+	verbose     bool
+	remoteFiles fileMap
+	localFiles  fileMap
+}
+
+type fileMap struct {
+	mutex sync.Mutex
+	files map[string]*fileInfo
+}
+
+func (fm *fileMap) add(key string, file *fileInfo) {
+	fm.mutex.Lock()
+	defer fm.mutex.Unlock()
+	fm.files[key] = file
+}
+
+func (fm *fileMap) get(key string) (*fileInfo, bool) {
+	fm.mutex.Lock()
+	defer fm.mutex.Unlock()
+	if file, ok := fm.files[key]; ok {
+		return file, true
+	}
+	return nil, false
+}
+
+// fileInfo represents a file in the filemanager.
+type fileInfo struct {
+	File
+}
+
+// NewFileManager returns a new filemanager.
+func NewFileManager() *Manager {
+	return &Manager{}
+}
+
+// WithClient sets the client
+func (fm *Manager) WithClient(client *Client) *Manager {
+	fm.client = client
+	return fm
+}
+
+// WithDelete sets the delete flag
+func (m *Manager) WithDelete(del bool) *Manager {
+	m.del = del
+	return m
+}
+
+// WithDryRun sets the dryrun flag
+func (m *Manager) WithDryRun(dryrun bool) *Manager {
+	m.dryrun = dryrun
+	return m
+}
+
+func (m *Manager) WithVerbose(verbose bool) *Manager {
+	m.verbose = verbose
+	return m
+}
+
+// Sync synchronizes the local directory with the filemanager.
+func (m *Manager) Sync(ctx context.Context, source, dest string) error {
+
+	if err := m.snapshotRemote(ctx, dest); err != nil {
+		return err
+	}
+
+	if err := m.snapshotLocal(source); err != nil {
+		return err
+	}
+
+	updates, err := m.filterUploads()
+
+	if err != nil {
+		return err
+	}
+
+	for _, uploadFile := range updates.files {
+		if uploadFile.IsDir {
+			continue
+		}
+
+		if err := m.upload(ctx, uploadFile, source, dest); err != nil {
+			return err
+		}
+	}
+
+	if m.del {
+		deletes := m.filterDeletes()
+		for _, remotePath := range deletes {
+			if err := m.deleteRemote(remotePath); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Purge deletes all files in the given path.
+func (m *Manager) Purge(ctx context.Context, remotePath string) error {
+
+	if err := m.snapshotRemote(ctx, remotePath); err != nil {
+		return err
+	}
+
+	deletes := m.filterDeletes()
```

Comment:
I don't think this is correct, as `m.filterDeletes()` consults local file system snapshot to check if files exists or not, but that's not something I would expect from this method. I would expect that give a path, every file and directory under that path will be removed.